include/swift/SILOptimizer/Utils/ScopeOptUtils.h

```diff
@@ -0,0 +1,284 @@
+  bool isValid() const { return scope && value; }
+
+  /// Returns true if this scoped value has a cleanup associated with it that
+  /// can be invalidated.
+  ///
+  /// Example: Result of SILOptScope::copyValue would return true, but result of
+  /// a SILOptScope::borrowValue would return false since we do not allow for
+  /// borrows to be cancelled.
+  bool hasInvalidatableCleanup() const {
+    assert(isValid());
+    return bool(scopeIndex);
+  }
+
+  /// Operator to access internal SILValue.
+  SILValue operator*() const {
+    assert(isValid());
+    return value;
+  }
+
+  /// Operator to call methods on internal SILValue.
+  SILValue operator->() const {
+    assert(isValid());
+    return value;
+  }
+
+  /// If this scope has a scoped index, invalidate the scoped value so that its
+  /// cleanup is not emitted at end of scope.
+  void invalidate() && {
+    assert(isValid());
+    // If we have an invalidated cleanup, do the invalidation.
+    if (hasInvalidatableCleanup()) {
+      scope->invalidateValue(std::move(*this));
+    }
+    // No matter what we do, make it so that if this object is used again, we
+    // get an assert.
+    scope = nullptr;
+    scopeIndex.reset();
+  }
+};
+
+inline void SILOptScopeBase::invalidateValue(ScopedValue &&value) {
+  if (value.scopeIndex)
+    invalidateCleanup(*value.scopeIndex);
+}
+
+/// A scope like stack data structure that one can use to manage cleanups for
+/// SILValues using light weight ScopedValue wrappers. This is done by using the
+/// copyValue/borrowValue methods. These copy/borrow the value and then push a
+/// destroy_value/end_borrow cleanup and return a ScopedValue containing that
+/// the user can then manipulate the value without needing to worry about the
+/// cleanup. If the user then decides a specific ScopedValue should be able to
+/// be cancelled, we allow for the ScopedValue to be invalidated
+class SILOptScope : public SILOptScopeBase {
+public:
+  SILOptScope() {}
+
+  /// Create a destroy_value cleanup for \p value. Returns a scoped value for \p
+  /// value with a live cleanup.
+  ScopedValue pushDestroyValue(SILValue value) {
+    // We need to capture value by value, not by address.
+    auto func = std::function<void(SILInstruction *)>(
+        [innerValue = value](SILInstruction *insertPt) {
+          SILBuilderWithScope builder(insertPt);
+          builder.emitDestroyValue(RegularLocation::getAutoGeneratedLocation(),
+                                   innerValue);
+        });
+    return {*this, value, pushScope(func)};
+  }
+
+  /// Copy \p value at \p insertPt and push a destroy_value cleanup. Return the
+  /// new copied value as a ScopedValue with cleanup.
+  ScopedValue copyValue(SILInstruction *insertPt, SILValue value) {
+    SILValue copy = SILBuilderWithScope(insertPt).emitCopyValueOperation(
+        insertPt->getLoc(), value);
+    return pushDestroyValue(copy);
+  }
+
+  ScopedValue pushEndBorrow(SILValue value) {
```

Comment:
It's slightly questionable as to why you'd bother creating a `ScopedValue` when there's no cancelable cleanup, but seems fine. Presumably you could assert that the scope is still active when the `ScopedValue` is used, but I don't know if there's any intention to actually do that.

Reply:
the reason why I did this is that in SILGen we found it interesting to be able to write APIs against ManagedValue that both had a cleanup/did not have a cleanup without having to change the currency type and thus have multiple APIs.